script/transform-approved-books-data.ts

```diff
@@ -0,0 +1,160 @@
+  return typeof something.repo === 'string' && isApprovedBooksData(something);
+};
+
+interface ApprovedVersion {
+  content_version: string;
+  min_code_version: string;
+}
+
+const isApprovedVersion = (something: any): something is ApprovedVersion => {
+  return typeof something.content_version === 'string' && typeof something.min_code_version === 'string';
+};
+
+interface ApprovedVersionCollection extends ApprovedVersion {
+  collection_id: string;
+}
+
+const isApprovedVersionCollection = (something: any): something is ApprovedVersionCollection => {
+  return typeof something.collection_id === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedVersionRepo extends ApprovedVersion {
+  repo: string;
+}
+
+const isApprovedVersionRepo = (something: any): something is ApprovedVersionRepo => {
+  return typeof something.repo === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedBooksAndVersions {
+  approved_books: Array<ApprovedCollection | ApprovedRepo>;
+  approved_versions: Array<ApprovedVersionCollection | ApprovedVersionRepo>;
+}
+
+const isApprovedBooksAndVersions = (something: any): something is ApprovedBooksAndVersions => {
+  return Array.isArray(something.approved_books)
+    && something.approved_books.every(
+      (element: any) => isApprovedCollection(element) || isApprovedRepo(element))
+    && something.approved_versions.every(
+      (element: any) => isApprovedVersionCollection(element) || isApprovedVersionRepo(element));
+};
+
+const matchRepoVersion = (repoData: ApprovedRepo) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionRepo => {
+  return isApprovedVersionRepo(versionData) && versionData.repo === repoData.repo;
+};
+
+const matchCollectionVersion = (collectionData: ApprovedCollection) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionCollection => {
+  return isApprovedVersionCollection(versionData) && versionData.collection_id === collectionData.collection_id;
+};
+
+const getDesiredVersion = (
+  approvedVersions: Array<ApprovedVersionCollection | ApprovedVersionRepo>,
+  colOrRepo: ApprovedRepo | ApprovedCollection
+): string | undefined => {
+  const isCollection = isApprovedCollection(colOrRepo);
+  const versions = approvedVersions
+    // only versions for current repo
+    .filter(isApprovedRepo(colOrRepo) ? matchRepoVersion(colOrRepo) : matchCollectionVersion(colOrRepo))
+    .map((versionData) => {
+      // books from collections have versions starting with 1.
+      return isCollection ? versionData.content_version.slice(2) : versionData.content_version;
+    })
+    // sort and revert so results are descending
+    .sort()
+    .reverse();
+
+  // in general we want the most recent one, in the future we'll have to filter this by which ones
+  // are available for the recipe code version REX supports
+  return versions[0];
```

Comment:
This is something that we could do after emerging #1133 and #1134 , right?

Reply:
@TomWoodward I've added the archive code version to the filter but I'm not sure if there is still a possibility that we will get more than one version matching the archive code version?

Reply:
there could be more than one version, we want the biggest version that matches our criteria